src/goto-instrument/synthesizer/expr_enumerator.cpp

```diff
@@ -0,0 +1,375 @@
+Module: Enumerator Interface
+Author: Qinheping Hu
+\*******************************************************************/
+
+#include "expr_enumerator.h"
+
+#include <util/simplify_expr.h>
+
+#include <langapi/language_util.h>
+
+expr_sett leaf_enumeratort::enumerate(const std::size_t size) const
+{
+  // Size of leaf expressions must be 1.
+  if(size != 1)
+    return {};
+
+  return leaf_exprs;
+}
+
+expr_sett non_leaf_enumeratort::enumerate(const std::size_t size) const
+{
+  expr_sett result;
+
+  // Enumerate nothing when `size` is too small to be partitioned.
+  if(size - 1 < arity)
+    return result;
+
+  // For every possible partition, set `size` of
+  // each sub-enumerator to be the corresponding component in the partition.
+  for(const auto &partition : get_partitions(size - 1, arity))
+  {
+    if(!is_good_partition(partition))
+      continue;
+
+    // Compute the Cartesian product as result.
+    for(const auto &product_tuple : cartesian_product_of_enumerators(
+          sub_enumerators,
+          sub_enumerators.begin(),
+          partition,
+          partition.begin()))
+    {
+      // Optimization: rule out equivalent expressions
+      // using certain equivalence class.
+      // Keep only representation tuple of each equivalence class.
+      if(is_equivalence_class_representation(product_tuple))
+        result.insert(simplify_expr(instantiate(product_tuple), ns));
+    }
+  }
+
+  return result;
+}
+
+std::set<expr_listt> non_leaf_enumeratort::cartesian_product_of_enumerators(
+  const enumeratorst &enumerators,
+  const enumeratorst::const_iterator &it_enumerators,
+  const partitiont &partition,
+  const partitiont::const_iterator &it_partition) const
+{
+  INVARIANT(
+    std::distance(it_enumerators, enumerators.end()) ==
+      std::distance(it_partition, partition.end()),
+    "Partition should have the same size as enumerators.");
+
+  std::set<expr_listt> result;
+
+  if(std::next(it_enumerators) == enumerators.end())
+  {
+    /// Current enumerator is the last enumerator.
+    /// Add all expressions enumerated by `it_enumerators` to `result`.
+    for(const auto &e : enumerators.back()->enumerate(*it_partition))
+    {
+      result.insert({e});
+    }
+  }
+  else
+  {
+    /// First compute the Cartesian product of enumerators after
+    /// `it_enumerators`. And then append the expressions enumerated by the
+    /// `it_enumerators` to every list in the Cartesian product.
+    for(const auto &sub_tuple : cartesian_product_of_enumerators(
+          enumerators,
+          std::next(it_enumerators),
+          partition,
+          std::next(it_partition)))
+    {
+      for(const auto &elem : (*it_enumerators)->enumerate(*it_partition))
+      {
+        expr_listt new_tuple(sub_tuple);
+        new_tuple.emplace_front(elem);
+        result.insert(new_tuple);
+      }
+    }
+  }
+  return result;
+}
+
+std::list<partitiont>
+get_partitions_64(const std::size_t n, const std::size_t k)
```

Comment:
Nit pick: not all supported platforms have 64-bit size_t types, so the name might be misleading?